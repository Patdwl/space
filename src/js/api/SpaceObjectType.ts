export enum SpaceObjectType {
  UNKNOWN = 0,
  PAYLOAD = 1,
  ROCKET_BODY = 2,
  DEBRIS = 3,
  SPECIAL = 4,
  RADAR_MEASUREMENT = 5,
  RADAR_TRACK = 6,
  RADAR_OBJECT = 7,
  BALLISTIC_MISSILE = 8,
  STAR = 9,
  INTERGOVERNMENTAL_ORGANIZATION = 10,
  SUBORBITAL_PAYLOAD_OPERATOR = 11,
  PAYLOAD_OWNER = 12,
  METEOROLOGICAL_ROCKET_LAUNCH_AGENCY_OR_MANUFACTURER = 13,
  PAYLOAD_MANUFACTURER = 14,
  LAUNCH_AGENCY = 15,
  LAUNCH_SITE = 16,
  LAUNCH_POSITION = 17,
  LAUNCH_FACILITY = 18,
  CONTROL_FACILITY = 19,
  GROUND_SENSOR_STATION = 20,
  OPTICAL = 21,
  MECHANICAL = 22,
  PHASED_ARRAY_RADAR = 23,
  OBSERVER = 24,
  BISTATIC_RADIO_TELESCOPE = 25,
  COUNTRY = 26,
  LAUNCH_VEHICLE_MANUFACTURER = 27,
  ENGINE_MANUFACTURER = 28,
}
