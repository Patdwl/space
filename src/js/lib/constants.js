"use strict";
exports.__esModule = true;
exports.STAR_DISTANCE = exports.RADIUS_OF_SUN = exports.GROUND_BUFFER_DISTANCE = exports.RADIUS_OF_EARTH = exports.cKmPerMs = exports.cKmPerSec = exports.cMPerSec = exports.MOON_SCALAR_DISTANCE = exports.RADIUS_OF_DRAW_MOON = exports.SUN_SCALAR_DISTANCE = exports.RADIUS_OF_DRAW_SUN = exports.PLANETARIUM_DIST = exports.MINUTES_PER_DAY = exports.MILLISECONDS_PER_DAY = exports.RAD2DEG = exports.DEG2RAD = exports.TAU = exports.PI = exports.ZOOM_EXP = void 0;
exports.ZOOM_EXP = 3;
exports.PI = Math.PI;
exports.TAU = 2 * Math.PI;
exports.DEG2RAD = exports.TAU / 360;
exports.RAD2DEG = 360 / exports.TAU;
// export const MILLISECONDS_PER_DAY = 1000 * 60 * 60 * 24;
// TODO: this is really MILLISECONDS_TO_DAYS
exports.MILLISECONDS_PER_DAY = 1.15741e-8;
exports.MINUTES_PER_DAY = 1440;
exports.PLANETARIUM_DIST = 3;
exports.RADIUS_OF_DRAW_SUN = 9000;
exports.SUN_SCALAR_DISTANCE = 250000;
exports.RADIUS_OF_DRAW_MOON = 4000;
exports.MOON_SCALAR_DISTANCE = 200000;
exports.cMPerSec = 299792458;
exports.cKmPerSec = 299792458 / 1000;
exports.cKmPerMs = 299792458 / 1000 / 1000;
exports.RADIUS_OF_EARTH = 6371; // Radius of Earth in kilometers
exports.GROUND_BUFFER_DISTANCE = 1; // Distance objects are placed above earth to avoid z-buffer fighting
exports.RADIUS_OF_SUN = 695700; // Radius of the Sun in kilometers
exports.STAR_DISTANCE = 250000; // Artificial Star Distance - Lower numberrReduces webgl depth buffer
