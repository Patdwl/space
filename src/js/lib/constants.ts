export const ZOOM_EXP = 3;
export const PI = Math.PI;
export const TAU = 2 * Math.PI;
export const DEG2RAD = TAU / 360;
export const RAD2DEG = 360 / TAU;
// export const MILLISECONDS_PER_DAY = 1000 * 60 * 60 * 24;
// TODO: this is really MILLISECONDS_TO_DAYS
export const MILLISECONDS_PER_DAY = 1.15741e-8;
export const MINUTES_PER_DAY = 1440;
export const PLANETARIUM_DIST = 3;
export const RADIUS_OF_DRAW_SUN = 9000;
export const SUN_SCALAR_DISTANCE = 250000;
export const RADIUS_OF_DRAW_MOON = 4000;
export const MOON_SCALAR_DISTANCE = 200000;
export const cMPerSec = 299792458;
export const cKmPerSec = 299792458 / 1000;
export const cKmPerMs = 299792458 / 1000 / 1000;
export const RADIUS_OF_EARTH = 6371; // Radius of Earth in kilometers
export const GROUND_BUFFER_DISTANCE = 1; // Distance objects are placed above earth to avoid z-buffer fighting
export const RADIUS_OF_SUN = 695700; // Radius of the Sun in kilometers
export const STAR_DISTANCE = 250000; // Artificial Star Distance - Lower numberrReduces webgl depth buffer
export const DISTANCE_TO_SUN = 149597870; // Distance from Earth to the Sun in kilometers
