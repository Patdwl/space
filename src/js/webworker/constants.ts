import * as Ootk from 'ootk';
import { SensorObjectCruncher } from '../api/keepTrackTypes';
import { SpaceObjectType } from '../api/SpaceObjectType';

// //////////////////////////////////////////////////////////////////////////////
// Typing
// //////////////////////////////////////////////////////////////////////////////

export interface SatCacheObject extends Ootk.SatelliteRecord {
  dec: number;
  ra: number;
  id?: number;
  isimp: number;
  apogee: number;
  isRadarData: any;
  static: boolean;
  marker: any;
  type: SpaceObjectType;
  lat: number;
  lon: number;
  alt: number;
  missile: any;
  active: any;
  altList: any;
  startTime: number;
  lastTime: any;
  latList: any;
  lonList: any;
  skip?: boolean;
}
export type PositionCruncherOutgoingMsg = {
  satInView?: Int8Array;
  satInSun?: Int8Array;
  sensorMarkerArray?: number[];
  satPos?: Float32Array;
  satVel?: Float32Array;
};
export type oneOrZero = 0 | 1;

// //////////////////////////////////////////////////////////////////////////////
// Code
// //////////////////////////////////////////////////////////////////////////////
export const defaultGd = {
  lat: <number | null>null,
  lon: 0,
  alt: 0,
};

export const emptySensor: SensorObjectCruncher = {
  observerGd: {
    lat: null,
    lon: 0,
    alt: 0,
  },
  alt: null,
  country: '',
  lat: null,
  lon: null,
  name: '',
  obsmaxaz: 0,
  obsmaxel: 0,
  obsmaxrange: 0,
  obsminaz: 0,
  obsminel: 0,
  obsminrange: 0,
  shortName: '',
  staticNum: 0,
  sun: '',
  volume: false,
  zoom: '',
};

type Radians = number;

export type RangeAzEl = {
  az: Radians;
  el: Radians;
  rng: number;
};

export interface PositionCruncherIncomingMsg {
  data: any;
}
