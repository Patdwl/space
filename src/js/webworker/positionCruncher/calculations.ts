import * as satellite from 'satellite.js'; // NOSONAR
import { SensorObjectCruncher } from '../../api/keepTrackTypes';
import { SpaceObjectType } from '../../api/SpaceObjectType';
import { DEG2RAD, MILLISECONDS_PER_DAY, PI, RAD2DEG } from '../../lib/constants';
import { A } from '../../lib/external/meuusjs';
import { jday } from '../../timeManager/transforms';
import { defaultGd, oneOrZero, RangeAzEl } from '../constants';

// TODO: create a way to determine if observerGd is using old satellite.js naming or lon,lat,alt
// so that we can import satMath.ts instead

export const lookAnglesToEcf = (azDeg: number, elDeg: number, rng: number, obsLat: number, obsLong: number, obsAlt: number) => {
  // site ecef in meters
  const geodeticCoords = {
    latitude: obsLat,
    longitude: obsLong,
    height: obsAlt,
  };

  const ecf = satellite.geodeticToEcf(geodeticCoords);

  // some needed calculations
  const slat = Math.sin(obsLat);
  const slon = Math.sin(obsLong);
  const clat = Math.cos(obsLat);
  const clon = Math.cos(obsLong);

  const azRad = DEG2RAD * azDeg;
  const elRad = DEG2RAD * elDeg;

  // az,el,range to sez convertion
  const south = -rng * Math.cos(elRad) * Math.cos(azRad);
  const east = rng * Math.cos(elRad) * Math.sin(azRad);
  const zenith = rng * Math.sin(elRad);

  const x = slat * clon * south + -slon * east + clat * clon * zenith + ecf.x;
  const y = slat * slon * south + clon * east + clat * slon * zenith + ecf.y;
  const z = -clat * south + slat * zenith + ecf.z;

  return { x, y, z };
};

/* Returns Current Propagation Time */
export const propTime = (dynamicOffsetEpoch: number, staticOffset: number, propRate: number) => {
  const now = new Date();
  const dynamicPropOffset = now.getTime() - dynamicOffsetEpoch;
  now.setTime(dynamicOffsetEpoch + staticOffset + dynamicPropOffset * propRate);
  return now;
};

export const checkSunExclusion = (sensor: SensorObjectCruncher, j: number, gmst: number, now: Date): [isSunExclusion: boolean, sunECI: { x: number; y: number; z: number }] => {
  const jdo = new A.JulianDay(j); // now
  const coord = A.EclCoordfromWgs84(0, 0, 0);
  const coord2 = A.EclCoordfromWgs84(sensor.observerGd.latitude * RAD2DEG, sensor.observerGd.longitude * RAD2DEG, sensor.observerGd.height);

  // AZ / EL Calculation
  const tp = <{ hz: { az: number; alt: number } }>(<unknown>A.Solar.topocentricPosition(jdo, coord, false));
  const tpRel = A.Solar.topocentricPosition(jdo, coord2, false);
  const sunAz = tp.hz.az * RAD2DEG + (180 % 360);
  const sunEl = (tp.hz.alt * RAD2DEG) % 360;
  const sunElRel = (tpRel.hz.alt * RAD2DEG) % 360;

  // Range Calculation
  const T = new A.JulianDay(A.JulianDay.dateToJD(now)).jdJ2000Century();
  let sunG = (A.Solar.meanAnomaly(T) * 180) / PI;
  sunG = sunG % 360.0;
  const sunR = 1.00014 - 0.01671 * Math.cos(sunG) - 0.00014 * Math.cos(2 * sunG);
  const sunRange = (sunR * 149597870700) / 1000; // au to km conversion

  // RAE to ECI
  const sunECI = satellite.ecfToEci(lookAnglesToEcf(sunAz, sunEl, sunRange, 0, 0, 0), gmst);
  return sensor.observerGd !== defaultGd && (sensor.type === SpaceObjectType.OPTICAL || sensor.type === SpaceObjectType.OBSERVER) && sunElRel > -6
    ? [true, sunECI]
    : [false, sunECI];
};

export const isInFov = (lookangles: RangeAzEl, sensor: SensorObjectCruncher): oneOrZero => {
  const azimuth = lookangles.azimuth * RAD2DEG;
  const elevation = lookangles.elevation * RAD2DEG;
  const rangeSat = lookangles.rangeSat;
  if (sensor.obsminaz > sensor.obsmaxaz) {
    if (
      ((azimuth >= sensor.obsminaz || azimuth <= sensor.obsmaxaz) &&
        elevation >= sensor.obsminel &&
        elevation <= sensor.obsmaxel &&
        rangeSat <= sensor.obsmaxrange &&
        rangeSat >= sensor.obsminrange) ||
      ((azimuth >= sensor.obsminaz2 || azimuth <= sensor.obsmaxaz2) &&
        elevation >= sensor.obsminel2 &&
        elevation <= sensor.obsmaxel2 &&
        rangeSat <= sensor.obsmaxrange2 &&
        rangeSat >= sensor.obsminrange2)
    ) {
      return 1;
    }
  } else {
    if (
      (azimuth >= sensor.obsminaz &&
        azimuth <= sensor.obsmaxaz &&
        elevation >= sensor.obsminel &&
        elevation <= sensor.obsmaxel &&
        rangeSat <= sensor.obsmaxrange &&
        rangeSat >= sensor.obsminrange) ||
      (azimuth >= sensor.obsminaz2 &&
        azimuth <= sensor.obsmaxaz2 &&
        elevation >= sensor.obsminel2 &&
        elevation <= sensor.obsmaxel2 &&
        rangeSat <= sensor.obsmaxrange2 &&
        rangeSat >= sensor.obsminrange2)
    ) {
      return 1;
    }
  }
  return 0;
};

export const setupTimeVariables = (dynamicOffsetEpoch, staticOffset, propRate, isSunlightView, isMultiSensor, sensor: SensorObjectCruncher) => {
  const now = propTime(dynamicOffsetEpoch, staticOffset, propRate);

  const j =
    jday(
      now.getUTCFullYear(),
      now.getUTCMonth() + 1, // Note, this function requires months in range 1-12.
      now.getUTCDate(),
      now.getUTCHours(),
      now.getUTCMinutes(),
      now.getUTCSeconds()
    ) +
    now.getUTCMilliseconds() * MILLISECONDS_PER_DAY;

  const gmst = satellite.gstime(j);

  let isSunExclusion = false;
  let sunEci = { x: 0, y: 0, z: 0 };
  if (isSunlightView && !isMultiSensor) {
    [isSunExclusion, sunEci] = checkSunExclusion(sensor, j, gmst, now);
  }

  const j2 =
    jday(
      now.getUTCFullYear(),
      now.getUTCMonth() + 1, // Note, this function requires months in range 1-12.
      now.getUTCDate(),
      now.getUTCHours(),
      now.getUTCMinutes(),
      now.getUTCSeconds() + 1
    ) +
    now.getUTCMilliseconds() * MILLISECONDS_PER_DAY;

  const gmstNext = satellite.gstime(j2);

  return {
    now,
    j,
    gmst,
    gmstNext,
    isSunExclusion,
    sunEci,
  };
};

export const createLatLonHei = (lat: number, lon: number, hei: number) => ({
  longitude: lon,
  latitude: lat,
  height: hei,
});

export const isInValidElevation = (lookangles: RangeAzEl, selectedSatFOV: number) => lookangles.elevation * RAD2DEG > 0 && 90 - lookangles.elevation * RAD2DEG < selectedSatFOV;

export const isSensorDeepSpace = (mSensor: SensorObjectCruncher[], sensor: SensorObjectCruncher): boolean => {
  // TODO: This should use the sensors max range instead of sensor type
  if (mSensor.length > 1 && sensor.type === SpaceObjectType.OPTICAL) return true;
  if (mSensor.length > 1 && sensor.type === SpaceObjectType.OBSERVER) return true;
  if (mSensor.length > 1 && sensor.type === SpaceObjectType.MECHANICAL) return true;
  return false;
};
