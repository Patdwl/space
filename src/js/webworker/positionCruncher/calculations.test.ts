import { SensorObjectCruncher } from '../../api/keepTrackTypes';
import { SpaceObjectType } from '../../api/SpaceObjectType';
import { DEG2RAD } from '../../lib/constants';
import * as calculations from './calculations';

const defaultSensor: SensorObjectCruncher = {
  observerGd: {
    lat: 0,
    lon: 0,
    alt: 0,
  },
  obsmaxaz: 0,
  obsmaxel: 0,
  obsminaz: 0,
  obsminel: 0,
  obsmaxrange: 0,
  obsminrange: 0,
  country: '',
  shortName: '',
  staticNum: 0,
  sun: '',
  lat: 0,
  lon: 0,
  alt: 0,
  name: '',
  volume: false,
  zoom: 'leo',
};

describe('calculations', () => {
  it('lookAnglesToEcf', () => {
    const azDeg = 0;
    const elDeg = 0;
    const rng = 0;
    const obsLat = 0;
    const obsLong = 0;
    const obsAlt = 0;
    const result = calculations.lookAnglesToEcf(azDeg, elDeg, rng, obsLat, obsLong, obsAlt);
    expect(result).toMatchSnapshot();
  });
  it('propTime', () => {
    const dynamicOffsetEpoch = 0;
    const staticOffset = 0;
    const propRate = 0;
    const now = new Date(2022, 0, 1);
    now.setUTCHours(0, 0, 0, 0);
    const result = calculations.propTime(dynamicOffsetEpoch, staticOffset, propRate);
    expect(result).toMatchSnapshot();
  });
  it('checkSunExclusion', () => {
    const j = 0;
    const gmst = 0;
    const now = new Date(2022, 0, 1);
    now.setUTCHours(0, 0, 0, 0);
    const result = calculations.checkSunExclusion(defaultSensor, j, gmst, now);
    expect(result).toMatchSnapshot();
  });

  describe('isInFov', () => {
    let isInFovSensor = { ...defaultSensor };
    let isInFovSensor2 = { ...defaultSensor };
    beforeEach(() => {
      isInFovSensor.obsminaz = -90;
      isInFovSensor.obsmaxaz = 90;
      isInFovSensor.obsminel = -90;
      isInFovSensor.obsmaxel = 90;
      isInFovSensor.obsminrange = 0;
      isInFovSensor.obsmaxrange = 100;

      isInFovSensor2.obsminaz = 0;
      isInFovSensor2.obsmaxaz = 0;
      isInFovSensor2.obsminel = 0;
      isInFovSensor2.obsmaxel = 0;
      isInFovSensor2.obsminrange = 0;
      isInFovSensor2.obsmaxrange = 0;
      isInFovSensor2.obsminaz2 = -90;
      isInFovSensor2.obsmaxaz2 = 90;
      isInFovSensor2.obsminel2 = -90;
      isInFovSensor2.obsmaxel2 = 90;
      isInFovSensor2.obsminrange2 = 0;
      isInFovSensor2.obsmaxrange2 = 100;
    });

    it('should return 1 if in fov', () => {
      const rae = {
        az: 0 * DEG2RAD,
        el: 0 * DEG2RAD,
        rng: 50,
      };
      const result = calculations.isInFov(rae, isInFovSensor);
      expect(result).toBe(1);
    });
    it('should return 0 if not in fov', () => {
      const rae = {
        az: -180 * DEG2RAD,
        el: -180 * DEG2RAD,
        rng: 200,
      };
      const result = calculations.isInFov(rae, isInFovSensor);
      expect(result).toBe(0);
    });
    it('should handle second sensor and in fov', () => {
      const rae = {
        az: 0 * DEG2RAD,
        el: 0 * DEG2RAD,
        rng: 50,
      };
      const result = calculations.isInFov(rae, isInFovSensor2);
      expect(result).toBe(1);
    });
    it('should handle second sensor, az crosses 0, and in fov', () => {
      const rae = {
        az: 0 * DEG2RAD,
        el: 0 * DEG2RAD,
        rng: 50,
      };
      isInFovSensor2.obsminaz = 350;
      isInFovSensor2.obsmaxaz = 10;
      isInFovSensor2.obsminaz2 = 350;
      isInFovSensor2.obsmaxaz2 = 10;
      const result = calculations.isInFov(rae, isInFovSensor2);
      expect(result).toBe(1);
    });
    it('should return 1 if in fov and az crosses 0', () => {
      const northSensor = { ...isInFovSensor };
      northSensor.obsminaz = 350;
      northSensor.obsmaxaz = 10;
      const rae = {
        az: 0 * DEG2RAD,
        el: 10 * DEG2RAD,
        rng: 50,
      };
      const result = calculations.isInFov(rae, northSensor);
      expect(result).toBe(1);
    });
    it('should return 0 if not in fov and az crosses 0', () => {
      const northSensor = { ...isInFovSensor };
      northSensor.obsminaz = 350;
      northSensor.obsmaxaz = 10;
      const rae = {
        az: 20 * DEG2RAD,
        el: -180 * DEG2RAD,
        rng: 200,
      };
      const result = calculations.isInFov(rae, northSensor);
      expect(result).toBe(0);
    });
  });
  describe('setupTimeVariables', () => {
    it('should return the correct time variables', () => {
      const now = new Date(2022, 0, 1);
      now.setUTCHours(0, 0, 0, 0);
      const result = calculations.setupTimeVariables(0, 0, 1, true, false, defaultSensor);
      expect(result).toMatchSnapshot();
    });
    it('should return the correct time variables', () => {
      const now = new Date(2022, 0, 1);
      now.setUTCHours(0, 0, 0, 0);
      const result = calculations.setupTimeVariables(0, 0, 1, false, false, defaultSensor);
      expect(result).toMatchSnapshot();
    });
  });

  describe('isSensorDeepSpace', () => {
    it('should return true if deep space', () => {
      const sensor = { ...defaultSensor };
      sensor.type = SpaceObjectType.OPTICAL;
      let result = calculations.isSensorDeepSpace([sensor, sensor], sensor);
      expect(result).toBe(true);

      sensor.type = SpaceObjectType.OBSERVER;
      result = calculations.isSensorDeepSpace([sensor, sensor], sensor);
      expect(result).toBe(true);

      sensor.type = SpaceObjectType.MECHANICAL;
      result = calculations.isSensorDeepSpace([sensor, sensor], sensor);
      expect(result).toBe(true);
    });

    it('should return false if not deep space', () => {
      const sensor = { ...defaultSensor };
      sensor.type = SpaceObjectType.PHASED_ARRAY_RADAR;
      let result = calculations.isSensorDeepSpace([sensor, sensor], sensor);
      expect(result).toBe(false);
      result = calculations.isSensorDeepSpace([sensor], sensor);
      expect(result).toBe(false);
      sensor.type = SpaceObjectType.MECHANICAL;
      result = calculations.isSensorDeepSpace([sensor], sensor);
      expect(result).toBe(false);
    });
  });
});
