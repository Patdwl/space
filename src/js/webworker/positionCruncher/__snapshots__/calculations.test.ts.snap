// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculations checkSunExclusion 1`] = `
Array [
  false,
  Object {
    "x": 137362374.578857,
    "y": -777123.1883858797,
    "z": -54884465.865863204,
  },
]
`;

exports[`calculations lookAnglesToEcf 1`] = `
Object {
  "x": 6378.137,
  "y": 0,
  "z": 0,
}
`;

exports[`calculations propTime 1`] = `1970-01-01T00:00:00.000Z`;

exports[`calculations setupTimeVariables should return the correct time variables 1`] = `
Object {
  "gmst": 1.7563256805284624,
  "gmstNext": 1.7563986012888932,
  "isSunExclusion": false,
  "j": 2459580.5,
  "now": 2022-01-01T00:00:00.000Z,
  "sunEci": Object {
    "x": 27051215.157581575,
    "y": -133418686.76125598,
    "z": -57840536.30867104,
  },
}
`;

exports[`calculations setupTimeVariables should return the correct time variables 2`] = `
Object {
  "gmst": 1.7563256805284624,
  "gmstNext": 1.7563986012888932,
  "isSunExclusion": false,
  "j": 2459580.5,
  "now": 2022-01-01T00:00:00.000Z,
  "sunEci": Object {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;
